flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCInputFormat.java

```diff
@@ -127,7 +134,7 @@ public JDBCInputFormat() {
 
 	@Override
 	public RowTypeInfo getProducedType() {
-		return rowTypeInfo;
+		return (RowTypeInfo) fromDataTypeToLegacyInfo(fromLogicalToDataType(rowType));
```

Comment:
changed it because it's easier to check data types when converting a jdbc row to flink row.

would you recommend revert it given I've changed JdbcTypeUtil.normalizeTableSchema() in the latest commit I pushed?

Reply:
As far as I can see, the rowType is used for `JdbcDialect#setRow`. IMO, `JdbcDialect` shoudn't be mixed in a runtime class. It's responsibility is to construct a correct runtime class based on the dialect. So I would suggest that `JdbcDialect#getJdbcRowConverter(RowType)` and the `JdbcInputFormat` accepts a `JdbcRowConverter`. The runtime converters should already be in place that simply perform conversion without going through big switch/case statements for types. 

In this way, we can still keep `rowTypeInfo` as the constructor parameter and don't need `rowType`. 

```java
interface JdbcRowConverter {
  Row convertToRow(ResultSet resultSet, Row reuse);
}
```

Reply:
I think 1) keeping rowTypeInfo v.s. rowType and 2) have row converter as runtime class, are orthogonal, right?

I agree keeping dialects out of runtime sounds like a better idea

Reply:
Yes. They are orthogonal. The reason we should use `TypeInformation` instead of `RowType` is `InputFormat` only works with `TypeInformation` and `RowType` is a concept of table world. 

The community is trying to reuse runtime implementation as much as possible. So a better design is that `JDBCInputFormat` accepts result `TypeInformation` and `JDBCConverter` as constructor parameters. So that, we can easily support generic type for DataStream users. We already refactored JDBC sink function to support generic type (see `JdbcBatchingOutputFormat`), and will refactor JDBC source soon.